client/beefy/src/round_tests.rs

```diff
@@ -0,0 +1,153 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use sc_network_test::Block;
+use sp_core::H256;
+use sp_runtime::traits::NumberFor;
+
+use beefy_primitives::{crypto::Public, ValidatorSet};
+
+use super::Rounds;
+use crate::keystore::tests::Keyring;
+
+#[test]
+fn new_rounds() {
+	sp_tracing::try_init_simple();
+
+	let rounds = Rounds::<H256, NumberFor<Block>>::new(ValidatorSet::<Public>::empty());
+
+	assert_eq!(0, rounds.validator_set_id());
+	assert!(rounds.validators().is_empty());
+
+	let validators = ValidatorSet::<Public> {
+		validators: vec![Keyring::Alice.public(), Keyring::Bob.public(), Keyring::Charlie.public()],
+		id: 42,
+	};
+
+	let rounds = Rounds::<H256, NumberFor<Block>>::new(validators);
+
+	assert_eq!(42, rounds.validator_set_id());
+
+	assert_eq!(
+		vec![Keyring::Alice.public(), Keyring::Bob.public(), Keyring::Charlie.public()],
+		rounds.validators()
+	);
+}
+
+#[test]
+fn add_vote() {
+	sp_tracing::try_init_simple();
+
+	let validators = ValidatorSet::<Public> {
+		validators: vec![Keyring::Alice.public(), Keyring::Bob.public(), Keyring::Charlie.public()],
+		id: Default::default(),
+	};
+
+	let mut rounds = Rounds::<H256, NumberFor<Block>>::new(validators);
+
+	assert!(rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Alice.public(), Keyring::Alice.sign(b"I am committed"))
+	));
+
+	assert!(!rounds.is_done(&(H256::from_low_u64_le(1), 1)));
+
+	// invalid vote
+	assert!(!rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Dave.public(), Keyring::Dave.sign(b"I am committed"))
+	));
+
+	assert!(!rounds.is_done(&(H256::from_low_u64_le(1), 1)));
+
+	assert!(rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Bob.public(), Keyring::Bob.sign(b"I am committed"))
+	));
+
+	assert!(!rounds.is_done(&(H256::from_low_u64_le(1), 1)));
+
+	assert!(rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Charlie.public(), Keyring::Charlie.sign(b"I am committed"))
+	));
+
+	assert!(rounds.is_done(&(H256::from_low_u64_le(1), 1)));
+}
+
+#[test]
+fn drop() {
+	sp_tracing::try_init_simple();
+
+	let validators = ValidatorSet::<Public> {
+		validators: vec![Keyring::Alice.public(), Keyring::Bob.public(), Keyring::Charlie.public()],
+		id: Default::default(),
+	};
+
+	let mut rounds = Rounds::<H256, NumberFor<Block>>::new(validators);
+
+	// round 1
+	rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Alice.public(), Keyring::Alice.sign(b"I am committed")),
+	);
+	rounds.add_vote(
+		(H256::from_low_u64_le(1), 1),
+		(Keyring::Bob.public(), Keyring::Bob.sign(b"I am committed")),
+	);
+
+	// round 2
+	rounds.add_vote(
+		(H256::from_low_u64_le(2), 2),
+		(Keyring::Alice.public(), Keyring::Alice.sign(b"I am again committed")),
```

Comment:
I find it a bit weird that `Rounds` component is responsible for verifying validator set, but not verifying signatures at all. I.e. here the payload changes and actually we could put any kind of gibberish and it would be totally fine.

Also it means that in `GossipValidator::validate` we are going to waste a lot of time verifying signatures without actually even looking if the validators are part of the validator set.

I.e. what we have:
1. Verification of the vote is arbitrarily split between two components (first signature, later validator set membership)
2. We validate things in the opposite order we should be validating them: i.e. we first check signature (expensive) and only then we check membership in some other component (cheap).

Note that this also means we will keep gossiping this useless votes. 

I'm totally aware that this behavior is not introduced in that particular PR, but just realised that while reviewing the tests.

Reply:
Review comment should be addressed as part of [Provide well-formed ValidatorSet](https://github.com/paritytech/grandpa-bridge-gadget/issues/307)